packages/waas/src/auth.ts

```diff
@@ -171,10 +184,90 @@ export class SequenceWaaS {
   ) {
     this.config = defaultArgsOrFail(config)
     this.waas = new SequenceWaaSBase({ network: 1, ...config }, this.store, this.cryptoBackend, this.secureStoreBackend)
-    this.client = new WaasAuthenticator(this.config.rpcServer, this.fetch.bind(this))
+    this.client = new WaasAuthenticator(this.config.rpcServer, this._fetch)
+
     this.deviceName = new StoreObj(this.store, '@0xsequence.waas.auth.deviceName', undefined)
   }
 
+  _fetch = (input: RequestInfo, init?: RequestInit): Promise<Response> => {
+    if (this.signatureVerificationFailed) {
+      throw new Error('Signature verification failed')
+    }
+
+    if (this.cryptoBackend && init?.headers) {
+      const headers: { [key: string]: any } = {}
+
+      headers['Accept-Signature'] = 'sig=();alg="rsa-v1_5-sha256"'
+
+      init!.headers = { ...init!.headers, ...headers }
+    }
+
+    const response = fetch(input, init)
+
+    if (this.cryptoBackend) {
+      response.then(async r => {
+        try {
+          const clone = r.clone()
+          const responseBodyText = await clone.text()
+
+          const contentDigest = r.headers.get('Content-Digest')
+          const signatureInput = r.headers.get('Signature-Input')
+          const signature = r.headers.get('Signature')
+
+          if (!contentDigest) {
+            throw new Error('Content-Digest header not set')
+          }
+          if (!signatureInput) {
+            throw new Error('Signature-Input header not set')
+          }
+          if (!signature) {
+            throw new Error('Signature header not set')
+          }
+
+          const contentDigestSha = contentDigest.match(':(.*):')?.[1]
+
+          if (!contentDigestSha) {
+            throw new Error('Content digest not found')
+          }
+
+          const responseBodyTextUint8Array = new TextEncoder().encode(responseBodyText)
+          const responseBodyTextDigest = await this.cryptoBackend!.digest('SHA-256', responseBodyTextUint8Array)
+          const base64EncodedDigest = btoa(String.fromCharCode(...responseBodyTextDigest))
+
+          if (contentDigestSha !== base64EncodedDigest) {
+            throw new Error('Digest mismatch')
+          }
+
+          const message = `"content-digest": ${contentDigest}\n"@signature-params": ${signatureInput.substring(4)}`
```

Comment:
Let's add a comment making it clear that "we're removing the first 4 characters to trim the `sig=` prefix"